File: node/be/bot/steps/bot.steps.js
'use strict';

const { Given, When, Then, Before } = require('@cucumber/cucumber');
const { Bot } = require('../client/bot');
const { RiskEngine } = require('../risk/engine');
const { Reverted, ChainUnreachable } = require('../../contract/chain/perpdex');
const { rpc } = require('../../contract/chain/ethcall');
const U = require('../../contract/chain/units');

const j = (v) => JSON.stringify(v, (k, x) => (typeof x === 'bigint' ? String(x) : x));

/**
 * Steps for features/be-bot.feature.
 *
 * The risk-gate scenarios build a RiskEngine directly with the deployed market
 * specs and never touch the chain, so they carry no @perpdex tag and the
 * contract tier's snapshot/revert hooks skip them.
 *
 * The chain scenarios are @perpdex, so those hooks give them a fresh venue and
 * a revert afterwards; the bot drives account `alice`, and reuses the contract
 * tier's Givens for counterparties ("a funded trader bob ...").
 */

// The three markets' parameters as the deploy script sets them. Fed to the
// pure risk engine so a unit scenario needs no chain, yet stays true to the venue.
const SPECS = {
  BTC: { tick: U.price('0.01'), step: U.size('0.0001'), minNotional: U.usd('10'), maxLeverage: 50n },
  ETH: { tick: U.price('0.01'), step: U.size('0.001'), minNotional: U.usd('10'), maxLeverage: 50n },
  SOL: { tick: U.price('0.001'), step: U.size('0.01'), minNotional: U.usd('10'), maxLeverage: 20n },
};

Before({ tags: '@bot' }, function () {
  this.risk = null;      // a bare RiskEngine, for the pure scenarios
  this.bot = null;       // a Bot over the chain, for the @perpdex scenarios
  this.plan = null;      // the last risk decision
  this.botLast = null;   // the last bot.place outcome (sent / decision / revert)
});

// ---------------------------------------------------------------- risk-only bot (pure)

Given('a risk-only bot with max order size {float} and max position {float}', function (maxOrder, maxPos) {
  this.risk = new RiskEngine({ specs: SPECS, maxOrderSize: U.size(String(maxOrder)), maxPositionSize: U.size(String(maxPos)) });
});
Given('the kill switch is engaged', function () { this.risk.setKillSwitch(true); });
Given('dry run is engaged', function () { this.risk.dryRun = true; });

function planOrder(world, over, ctx = {}) {
  const o = { market: 'ETH', side: 'Buy', size: U.size('1'), price: U.price('2500'), orderType: 'Limit', ...over };
  world.plan = world.risk.plan(o, ctx);
}

When('it plans a buy of {float} {word} at {float}', function (size, sym, price) {
  planOrder(this, { market: sym, size: U.size(String(size)), price: U.price(String(price)) });
});
When('it plans a buy of {float} {word} at {float} with user order id {int}', function (size, sym, price, uid) {
  planOrder(this, { market: sym, size: U.size(String(size)), price: U.price(String(price)), userOrderId: uid });
});
When('it plans a buy of {float} {word} at {float} against an existing position of {float} {word}', function (size, sym, price, pos, _u) {
  planOrder(this, { market: sym, size: U.size(String(size)), price: U.price(String(price)) }, { positionSize: U.size(String(pos)) });
});
When('it plans a reduce-only sell of {float} {word} at {float} against an existing position of {float} {word}', function (size, sym, price, pos, _u) {
  planOrder(this, { market: sym, side: 'Sell', size: U.size(String(size)), price: U.price(String(price)), reduceOnly: true, userOrderId: 999 }, { positionSize: U.size(String(pos)) });
});

// These read the pure risk decision (this.plan) or, for a chain bot that
// decided before sending, its last outcome (this.botLast.decision).
function decision(world) { return world.plan || (world.botLast && world.botLast.decision) || null; }
Then('the order is cleared to send', function () {
  const d = decision(this); this.observe('order cleared to send', () => ({ passed: !!d && d.ok && d.action === 'send', detail: j(d) }));
});
Then('the order is cleared as a dry run', function () {
  const d = decision(this); this.observe('order cleared as dry run', () => ({ passed: !!d && d.ok && d.action === 'dry-run', detail: j(d) }));
});
Then('the order is a duplicate', function () {
  const d = decision(this); this.observe('order is a duplicate', () => ({ passed: !!d && d.ok && d.action === 'duplicate', detail: j(d) }));
});
Then('the order is refused with reason {string}', function (reason) {
  // The decision is in `this.plan` for a pure risk scenario, or in the chain
  // bot's last outcome when the bot itself refused before sending.
  const d = this.plan || (this.botLast && this.botLast.decision);
  this.observe('order refused: ' + reason, () => ({ passed: !!d && d.ok === false && d.reason === reason, detail: j(d) }));
});
Then('the planned size is {float} {word}', function (size, _sym) {
  this.observe('planned size', () => ({ passed: this.plan.rounded && this.plan.rounded.size === U.size(String(size)), detail: this.plan.rounded ? U.showSize(this.plan.rounded.size) : 'no rounded' }));
});
Then('the planned price is {float}', function (price) {
  this.observe('planned price', () => ({ passed: this.plan.rounded && this.plan.rounded.price === U.price(String(price)), detail: this.plan.rounded ? U.showPrice(this.plan.rounded.price) : 'no rounded' }));
});

Then('a transient failure is retried up to the limit', async function () {
  const bot = new Bot({ dex: { url: '' }, account: '0x0', risk: this.risk });
  let calls = 0;
  try { await bot.withRetry('t', async () => { calls += 1; throw new ChainUnreachable('down'); }, { max: 3 }); }
  catch (err) { if (!(err instanceof ChainUnreachable)) throw err; }
  this.check('a transient is retried up to the limit', calls === 3, 'called ' + calls + ' times, expected 3');
});
Then('a revert is surfaced on the first attempt', async function () {
  const bot = new Bot({ dex: { url: '' }, account: '0x0', risk: this.risk });
  let calls = 0; let threw = false;
  try { await bot.withRetry('t', async () => { calls += 1; throw new Reverted({ name: 'PostOnlyWouldCross', args: [] }, 'x'); }, { max: 3 }); }
  catch (err) { if (!(err instanceof Reverted)) throw err; threw = true; }
  this.check('a revert is surfaced on the first attempt', threw && calls === 1, 'threw=' + threw + ', called ' + calls + ' times, expected 1');
});
Then('the bot\'s record contains no secret', function () {
  const bot = new Bot({ dex: { url: '' }, account: '0x0', risk: this.risk, log: ['funded account with $100000', 'plan Buy 1 ETH -> send', 'placed order 3'] });
  this.observe('no secret in the bot record', () => ({ passed: bot.leakedSecret() === null, detail: bot.leakedSecret() || 'clean' }));
});

// ---------------------------------------------------------------- bot over the chain

const CAPS = { maxOrderSize: U.size('1000'), maxPositionSize: U.size('1000') };

async function makeBot(world, usd, opts = {}) {
  if (world.sourceError) return;
  try {
    world.bot = await Bot.fromChain(world.dex, world.who.alice, { ...CAPS, ...opts });
    await world.bot.fund(usd);
    world.noted.nonce0 = null;
  } catch (err) {
    if (err instanceof ChainUnreachable) { world.sourceError = err.message; return; }
    throw err;
  }
}

Given('a bot funded with ${float}', { timeout: 60_000 }, async function (usd) { await makeBot(this, usd); });
Given('a dry-run bot funded with ${float}', { timeout: 60_000 }, async function (usd) { await makeBot(this, usd, { dryRun: true }); });
Given('the bot\'s kill switch is engaged', function () { if (this.bot) this.bot.risk.setKillSwitch(true); });

const SIDE = { buy: 'Buy', sell: 'Sell' };

/** Place through the bot, capturing a revert instead of throwing. */
async function botPlace(world, spec) {
  if (world.sourceError || !world.bot) { world.botLast = { skipped: true }; return; }
  try {
    world.botLast = await world.bot.place(spec);
    if (world.botLast.sent && world.botLast.orderId !== undefined) world.orders.bot = world.botLast.orderId;
  } catch (err) {
    if (err instanceof Reverted) { world.botLast = { sent: false, revert: err }; return; }
    if (err instanceof ChainUnreachable) { world.sourceError = err.message; world.botLast = { skipped: true }; return; }
    throw err;
  }
}

const placeStep = (kind) => async function (side, size, sym, price) {
  await botPlace(this, { sym, side: SIDE[side], size, price, orderType: kind.type, tif: kind.tif });
};
// Given/When share one registry in cucumber-js, so each text is defined once.
When('the bot places a limit {word} of {float} {word} at {float}', { timeout: 30_000 }, placeStep({ type: 'Limit', tif: 'GTC' }));
When('the bot places a post-only {word} of {float} {word} at {float}', { timeout: 30_000 }, placeStep({ type: 'Limit', tif: 'PostOnly' }));
When('the bot places an IOC {word} of {float} {word} at {float}', { timeout: 30_000 }, placeStep({ type: 'Limit', tif: 'IOC' }));
When('the bot places a {word} of {float} {word} at {float}', { timeout: 30_000 }, async function (side, size, sym, price) {
  await botPlace(this, { sym, side: SIDE[side], size, price, orderType: 'Limit', tif: 'GTC' });
});
When('the bot places a reduce-only {word} of {float} {word} at {float}', { timeout: 30_000 }, async function (side, size, sym, price) {
  await botPlace(this, { sym, side: SIDE[side], size, price, orderType: 'Limit', tif: 'GTC', reduceOnly: true });
});
When('the bot buys {float} {word} at market', { timeout: 30_000 }, async function (size, sym) {
  await botPlace(this, { sym, side: 'Buy', size, orderType: 'Market', tif: 'IOC' });
});

When('the bot cancels its order', { timeout: 30_000 }, async function () {
  if (!this.sourceError && this.bot) await this.bot.cancel(this.orders.bot).catch((e) => { if (!(e instanceof ChainUnreachable)) throw e; else this.sourceError = e.message; });
});
When('the bot cancels all its orders', { timeout: 30_000 }, async function () {
  if (!this.sourceError && this.bot) await this.bot.cancelAll().catch((e) => { if (!(e instanceof ChainUnreachable)) throw e; else this.sourceError = e.message; });
});
When('the bot flattens {word}', { timeout: 30_000 }, async function (sym) {
  if (!this.sourceError && this.bot) this.noted.flat = await this.bot.flatten(sym).catch((e) => { if (!(e instanceof ChainUnreachable)) throw e; else { this.sourceError = e.message; return null; } });
});

// ---------------------------------------------------------------- bot Thens (chain)

async function chk(world, description, fn) {
  if (world.sourceError) { world.unobservable(description, 'the source could not be reached -- ' + world.sourceError); return; }
  let r;
  try { r = await fn(); } catch (err) { if (err instanceof ChainUnreachable) { world.unobservable(description, err.message); return; } throw err; }
  world.check(description, r.passed, r.detail);
}

Then('the bot\'s order reverts with {word}', function (name) {
  this.observe('bot order reverts with ' + name, () => {
    if (!this.botLast || !this.botLast.revert) return { passed: false, detail: this.botLast && this.botLast.sent ? 'the order was sent, no revert' : 'no revert recorded (' + JSON.stringify(this.botLast) + ')' };
    return { passed: this.botLast.revert.error === name, detail: this.botLast.revert.message };
  });
});
Then('the bot tried the place once', function () {
  this.observe('place attempted once', () => { const a = (this.bot.attempts || []).filter((x) => x.op === 'place'); return { passed: a.length === 1 && a[0].tries === 1, detail: JSON.stringify(a) }; });
});
Then('the bot\'s account exists', async function () {
  await chk(this, 'bot account exists', async () => ({ passed: await this.dex.accountExists(this.who.alice), detail: 'accountExists false' }));
});
Then('the bot\'s vault balance is ${float}', async function (usd) {
  await chk(this, 'bot vault balance', async () => { const b = await this.dex.vaultBalance(this.who.alice); return { passed: b === U.usd(String(usd)), detail: '$' + U.showUsd(b) }; });
});
Then('the bot\'s order is Open', function () {
  this.observe('bot order Open', () => ({ passed: !!(this.botLast && this.botLast.order && this.botLast.order.statusName === 'Open'), detail: this.botLast && this.botLast.order ? this.botLast.order.statusName : JSON.stringify(this.botLast) }));
});
Then('the bot\'s order is Cancelled', async function () {
  await chk(this, 'bot order Cancelled', async () => { const o = await this.dex.getOrder(this.orders.bot); return { passed: o.statusName === 'Cancelled', detail: o.statusName }; });
});
Then('the bot\'s order size is {float} {word}', function (size, _u) {
  this.observe('bot order size', () => ({ passed: !!(this.botLast && this.botLast.order && this.botLast.order.size === U.size(String(size))), detail: this.botLast && this.botLast.order ? U.showSize(this.botLast.order.size) : 'no order' }));
});
Then('the bot\'s order price is {float}', function (price) {
  this.observe('bot order price', () => ({ passed: !!(this.botLast && this.botLast.order && this.botLast.order.price === U.price(String(price))), detail: this.botLast && this.botLast.order ? U.showPrice(this.botLast.order.price) : 'no order' }));
});
Then('the bot has {int} open orders', async function (n) {
  await chk(this, 'bot open orders ' + n, async () => { const a = await this.dex.getAccount(this.who.alice); return { passed: Number(a.openOrders) === n, detail: 'openOrders ' + a.openOrders }; });
});
Then('the bot is flat on {word}', async function (sym) {
  await chk(this, 'bot flat on ' + sym, () => ({ passed: !!(this.noted.flat && this.noted.flat.flat), detail: j(this.noted.flat) }));
});
Then('the bot\'s {word} position size is {float} {word}', async function (sym, size, _u) {
  const mid = { BTC: 0, ETH: 1, SOL: 2 }[sym];
  await chk(this, 'bot ' + sym + ' size', async () => { const p = await this.dex.getPosition(this.who.alice, mid); return { passed: p.size === U.size(String(size)), detail: U.showSize(p.size) }; });
});
// "the backstop's <sym> position size is ..." is already defined by the contract tier.
Then('the bot\'s {word} entry price is above 0', async function (sym) {
  const mid = { BTC: 0, ETH: 1, SOL: 2 }[sym];
  await chk(this, 'bot ' + sym + ' entry > 0', async () => { const p = await this.dex.getPosition(this.who.alice, mid); return { passed: p.entryPrice > 0n, detail: U.showPrice(p.entryPrice) }; });
});
Then('the bot\'s free collateral fell', async function () {
  await chk(this, 'bot free collateral fell', async () => { const f = await this.dex.freeCollateral(this.who.alice); const bal = await this.dex.vaultBalance(this.who.alice); return { passed: f < bal, detail: 'free $' + U.showUsd(f) + ' < balance $' + U.showUsd(bal) }; });
});
Then('the treasury\'s vault balance increased', async function () {
  await chk(this, 'treasury balance increased', async () => { const b = await this.dex.vaultBalance(this.who.treasury); return { passed: b > this.noted['bal.treasury'], detail: 'before ' + String(this.noted['bal.treasury']) + ', after ' + String(b) }; });
});
Then('the bot\'s fill was mined', function () {
  this.observe('fill mined', () => { const r = this.botLast && this.botLast.receipt && this.botLast.receipt.receipt; return { passed: !!r && r.status === '0x1' && !!r.blockNumber, detail: r ? 'status ' + r.status + ' block ' + Number(BigInt(r.blockNumber)) : 'no receipt' }; });
});
Then('the account nonce advanced by {int}', async function (n) {
  await chk(this, 'nonce advanced by ' + n, async () => {
    const now = Number(BigInt(await rpc('eth_getTransactionCount', [this.who.alice, 'latest'], this.dex.url)));
    const first = this.bot.nonces[0];
    const consecutive = this.bot.nonces.every((v, i) => i === 0 || v === this.bot.nonces[i - 1] + 1);
    return { passed: consecutive && now - first === n, detail: 'nonces ' + JSON.stringify(this.bot.nonces) + ', now ' + now };
  });
});
Then('the bot\'s {word} order clears the minimum notional', async function (sym) {
  const mid = { BTC: 0, ETH: 1, SOL: 2 }[sym];
  await chk(this, sym + ' order clears min notional', async () => {
    const o = this.botLast.order; const p = await this.dex.getMarketParams(mid);
    const notional = U.notional(o.size, o.price);
    return { passed: notional >= p.minNotional, detail: 'notional $' + U.showUsd(notional) + ' >= min $' + U.showUsd(p.minNotional) };
  });
});

File: python/be/bot/steps/bot_steps.py
"""Steps for be-bot.feature -- a plugin module. Mirror of node/be/bot/steps/bot.steps.js.

Risk-gate scenarios build a RiskEngine directly with the deployed market specs
and never touch the chain (no @perpdex tag, so the contract tier's hooks skip
them). Chain scenarios are @perpdex; the bot drives account `alice` and reuses
the contract tier's Givens for counterparties.
"""

from __future__ import annotations

import json

import pytest
from pytest_bdd import given, then, when

from be.bot.client.bot import Bot
from be.bot.risk.engine import RiskEngine
from be.contract.chain import units as U
from be.contract.chain.ethcall import ChainUnreachable, rpc
from be.contract.chain.perpdex import Reverted
from be.contract.steps.perpdex_steps import P, market_id

# The three markets' parameters as the deploy script sets them -- fed to the pure
# risk engine so a unit scenario needs no chain yet stays true to the venue.
SPECS = {
    "BTC": {"tick": U.price("0.01"), "step": U.size("0.0001"), "minNotional": U.usd("10"), "maxLeverage": 50},
    "ETH": {"tick": U.price("0.01"), "step": U.size("0.001"), "minNotional": U.usd("10"), "maxLeverage": 50},
    "SOL": {"tick": U.price("0.001"), "step": U.size("0.01"), "minNotional": U.usd("10"), "maxLeverage": 20},
}
SIDE = {"buy": "Buy", "sell": "Sell"}
MID = {"BTC": 0, "ETH": 1, "SOL": 2}
CAPS = {"max_order_size": U.size("1000"), "max_position_size": U.size("1000")}


def j(v) -> str:
    return json.dumps(v, default=str)


@pytest.fixture(autouse=True)
def bot_scenario(request, qa, perpdex_scenario):
    """Per-scenario bot slots. Depends on perpdex_scenario so the chain slots exist first."""
    if request.node.get_closest_marker("bot") is None:
        yield
        return
    qa.risk = None
    qa.plan = None
    qa.bot = None
    qa.bot_last = None
    yield


# ---------------------------------------------------------------- risk-only bot (pure)


@given(P("a risk-only bot with max order size {max_order:dec} and max position {max_pos:dec}"))
def risk_only_bot(qa, max_order, max_pos):
    qa.risk = RiskEngine(specs=SPECS, max_order_size=U.size(max_order), max_position_size=U.size(max_pos))


@given("the kill switch is engaged")
def kill_switch(qa):
    qa.risk.set_kill_switch(True)


@given("dry run is engaged")
def dry_run(qa):
    qa.risk.dry_run = True


def _plan(qa, over, ctx=None):
    o = {"market": "ETH", "side": "Buy", "size": U.size("1"), "price": U.price("2500"), "orderType": "Limit"}
    o.update(over)
    qa.plan = qa.risk.plan(o, ctx or {})


@when(P("it plans a buy of {size:dec} {sym:w} at {price:dec}"))
def plan_buy(qa, size, sym, price):
    _plan(qa, {"market": sym, "size": U.size(size), "price": U.price(price)})


@when(P("it plans a buy of {size:dec} {sym:w} at {price:dec} with user order id {uid:d}"))
def plan_buy_uid(qa, size, sym, price, uid):
    _plan(qa, {"market": sym, "size": U.size(size), "price": U.price(price), "userOrderId": uid})


@when(P("it plans a buy of {size:dec} {sym:w} at {price:dec} against an existing position of {pos:dec} {u:w}"))
def plan_buy_pos(qa, size, sym, price, pos, u):
    _plan(qa, {"market": sym, "size": U.size(size), "price": U.price(price)}, {"positionSize": U.size(pos)})


@when(P("it plans a reduce-only sell of {size:dec} {sym:w} at {price:dec} against an existing position of {pos:dec} {u:w}"))
def plan_reduce(qa, size, sym, price, pos, u):
    _plan(qa, {"market": sym, "side": "Sell", "size": U.size(size), "price": U.price(price), "reduceOnly": True, "userOrderId": 999}, {"positionSize": U.size(pos)})


def _decision(qa):
    return qa.plan or (qa.bot_last or {}).get("decision")


@then("the order is cleared to send")
def cleared_send(qa):
    d = _decision(qa)
    qa.observe("order cleared to send", lambda: (bool(d) and d["ok"] and d["action"] == "send", j(d)))


@then("the order is cleared as a dry run")
def cleared_dry_run(qa):
    d = _decision(qa)
    qa.observe("order cleared as dry run", lambda: (bool(d) and d["ok"] and d["action"] == "dry-run", j(d)))


@then("the order is a duplicate")
def is_duplicate(qa):
    d = _decision(qa)
    qa.observe("order is a duplicate", lambda: (bool(d) and d["ok"] and d["action"] == "duplicate", j(d)))


@then(P('the order is refused with reason "{reason}"'))
def refused(qa, reason):
    d = _decision(qa)
    qa.observe("order refused: " + reason, lambda: (bool(d) and d["ok"] is False and d["reason"] == reason, j(d)))


@then(P("the planned size is {size:dec} {u:w}"))
def planned_size(qa, size, u):
    qa.observe("planned size", lambda: (bool(qa.plan.get("rounded")) and qa.plan["rounded"]["size"] == U.size(size), U.show_size(qa.plan["rounded"]["size"]) if qa.plan.get("rounded") else "no rounded"))


@then(P("the planned price is {price:dec}"))
def planned_price(qa, price):
    qa.observe("planned price", lambda: (bool(qa.plan.get("rounded")) and qa.plan["rounded"]["price"] == U.price(price), U.show_price(qa.plan["rounded"]["price"]) if qa.plan.get("rounded") else "no rounded"))


@then("a transient failure is retried up to the limit")
def transient_retried(qa):
    bot = Bot(dex=type("D", (), {"url": ""})(), account="0x0", risk=qa.risk)
    calls = {"n": 0}

    def boom():
        calls["n"] += 1
        raise ChainUnreachable("down")
    try:
        bot.with_retry("t", boom, max_tries=3)
    except ChainUnreachable:
        pass
    qa.check("a transient is retried up to the limit", calls["n"] == 3, f"called {calls['n']} times, expected 3")


@then("a revert is surfaced on the first attempt")
def revert_not_retried(qa):
    bot = Bot(dex=type("D", (), {"url": ""})(), account="0x0", risk=qa.risk)
    calls = {"n": 0}
    threw = False

    def boom():
        calls["n"] += 1
        raise Reverted({"name": "PostOnlyWouldCross", "args": []}, "x")
    try:
        bot.with_retry("t", boom, max_tries=3)
    except Reverted:
        threw = True
    qa.check("a revert is surfaced on the first attempt", threw and calls["n"] == 1, f"threw={threw}, called {calls['n']} times, expected 1")


@then("the bot's record contains no secret")
def no_secret(qa):
    bot = Bot(dex=type("D", (), {"url": ""})(), account="0x0", risk=qa.risk, log=["funded account with $100000", "plan Buy 1 ETH -> send", "placed order 3"])
    qa.observe("no secret in the bot record", lambda: (bot.leaked_secret() is None, bot.leaked_secret() or "clean"))


# ---------------------------------------------------------------- bot over the chain


def make_bot(qa, usd, **opts):
    if qa.source_error:
        return
    try:
        qa.bot = Bot.from_chain(qa.dex, qa.who["alice"], CAPS["max_order_size"], CAPS["max_position_size"], **opts)
        qa.bot.fund(usd)
    except ChainUnreachable as err:
        qa.source_error = str(err)


@given(P("a bot funded with ${usd:dec}"))
def bot_funded(qa, usd):
    make_bot(qa, usd)


@given(P("a dry-run bot funded with ${usd:dec}"))
def dry_run_bot_funded(qa, usd):
    make_bot(qa, usd, dry_run=True)


@given("the bot's kill switch is engaged")
def bot_kill_switch(qa):
    if qa.bot:
        qa.bot.risk.set_kill_switch(True)


def bot_place(qa, **spec):
    if qa.source_error or not qa.bot:
        qa.bot_last = {"skipped": True}
        return
    try:
        qa.bot_last = qa.bot.place(**spec)
        if qa.bot_last.get("sent") and qa.bot_last.get("orderId") is not None:
            qa.orders["bot"] = qa.bot_last["orderId"]
    except Reverted as err:
        qa.bot_last = {"sent": False, "revert": err}
    except ChainUnreachable as err:
        qa.source_error = str(err)
        qa.bot_last = {"skipped": True}


@given(P("the bot places a limit {side:w} of {size:dec} {sym:w} at {price:dec}"))
@when(P("the bot places a limit {side:w} of {size:dec} {sym:w} at {price:dec}"))
def bot_places_limit(qa, side, size, sym, price):
    bot_place(qa, sym=sym, side=SIDE[side], size=size, price=price, order_type="Limit", tif="GTC")


@when(P("the bot places a post-only {side:w} of {size:dec} {sym:w} at {price:dec}"))
def bot_places_postonly(qa, side, size, sym, price):
    bot_place(qa, sym=sym, side=SIDE[side], size=size, price=price, order_type="Limit", tif="PostOnly")


@when(P("the bot places an IOC {side:w} of {size:dec} {sym:w} at {price:dec}"))
def bot_places_ioc(qa, side, size, sym, price):
    bot_place(qa, sym=sym, side=SIDE[side], size=size, price=price, order_type="Limit", tif="IOC")


@when(P("the bot places a {side:w} of {size:dec} {sym:w} at {price:dec}"))
def bot_places_plain(qa, side, size, sym, price):
    bot_place(qa, sym=sym, side=SIDE[side], size=size, price=price, order_type="Limit", tif="GTC")


@when(P("the bot places a reduce-only {side:w} of {size:dec} {sym:w} at {price:dec}"))
def bot_places_reduce(qa, side, size, sym, price):
    bot_place(qa, sym=sym, side=SIDE[side], size=size, price=price, order_type="Limit", tif="GTC", reduce_only=True)


@given(P("the bot buys {size:dec} {sym:w} at market"))
@when(P("the bot buys {size:dec} {sym:w} at market"))
def bot_buys_market(qa, size, sym):
    bot_place(qa, sym=sym, side="Buy", size=size, order_type="Market", tif="IOC")


def _guard(qa, fn):
    if qa.source_error or not qa.bot:
        return
    try:
        return fn()
    except ChainUnreachable as err:
        qa.source_error = str(err)


@when("the bot cancels its order")
def bot_cancels(qa):
    _guard(qa, lambda: qa.bot.cancel(qa.orders["bot"]))


@when("the bot cancels all its orders")
def bot_cancels_all(qa):
    _guard(qa, lambda: qa.bot.cancel_all())


@when(P("the bot flattens {sym:w}"))
def bot_flattens(qa, sym):
    qa.noted["flat"] = _guard(qa, lambda: qa.bot.flatten(sym))


# ---------------------------------------------------------------- bot Thens (chain)


def chk(qa, description, fn):
    if qa.source_error:
        qa.unobservable(description, "the source could not be reached -- " + qa.source_error)
        return
    try:
        passed, detail = fn()
    except ChainUnreachable as err:
        qa.unobservable(description, str(err))
        return
    qa.check(description, passed, detail)


@then(P("the bot's order reverts with {name:w}"))
def bot_reverts(qa, name):
    def ev():
        bl = qa.bot_last
        if not bl or not bl.get("revert"):
            return False, ("the order was sent, no revert" if bl and bl.get("sent") else "no revert recorded (" + j(bl) + ")")
        return bl["revert"].error == name, str(bl["revert"])
    qa.observe("bot order reverts with " + name, ev)


@then("the bot tried the place once")
def tried_once(qa):
    a = [x for x in (qa.bot.attempts or []) if x["op"] == "place"]
    qa.observe("place attempted once", lambda: (len(a) == 1 and a[0]["tries"] == 1, j(a)))


@then("the bot's account exists")
def bot_account_exists(qa):
    chk(qa, "bot account exists", lambda: (qa.dex.account_exists(qa.who["alice"]), "accountExists false"))


@then(P("the bot's vault balance is ${usd:dec}"))
def bot_vault_balance(qa, usd):
    chk(qa, "bot vault balance", lambda: ((lambda b: (b == U.usd(usd), "$" + U.show_usd(b)))(qa.dex.vault_balance(qa.who["alice"]))))


@then("the bot's order is Open")
def bot_order_open(qa):
    o = (qa.bot_last or {}).get("order")
    qa.observe("bot order Open", lambda: (bool(o) and o["statusName"] == "Open", o["statusName"] if o else j(qa.bot_last)))


@then("the bot's order is Cancelled")
def bot_order_cancelled(qa):
    chk(qa, "bot order Cancelled", lambda: ((lambda o: (o["statusName"] == "Cancelled", o["statusName"]))(qa.dex.get_order(qa.orders["bot"]))))


@then(P("the bot's order size is {size:dec} {u:w}"))
def bot_order_size(qa, size, u):
    o = (qa.bot_last or {}).get("order")
    qa.observe("bot order size", lambda: (bool(o) and o["size"] == U.size(size), U.show_size(o["size"]) if o else "no order"))


@then(P("the bot's order price is {price:dec}"))
def bot_order_price(qa, price):
    o = (qa.bot_last or {}).get("order")
    qa.observe("bot order price", lambda: (bool(o) and o["price"] == U.price(price), U.show_price(o["price"]) if o else "no order"))


@then(P("the bot has {n:d} open orders"))
def bot_open_orders(qa, n):
    chk(qa, f"bot open orders {n}", lambda: ((lambda a: (int(a["openOrders"]) == n, "openOrders " + str(a["openOrders"])))(qa.dex.get_account(qa.who["alice"]))))


@then(P("the bot is flat on {sym:w}"))
def bot_flat(qa, sym):
    chk(qa, "bot flat on " + sym, lambda: (bool(qa.noted.get("flat") and qa.noted["flat"]["flat"]), j(qa.noted.get("flat"))))


@then(P("the bot's {sym:w} position size is {size:dec} {u:w}"))
def bot_position_size(qa, sym, size, u):
    chk(qa, f"bot {sym} size", lambda: ((lambda p: (p["size"] == U.size(size), U.show_size(p["size"])))(qa.dex.get_position(qa.who["alice"], MID[sym]))))


@then(P("the bot's {sym:w} entry price is above 0"))
def bot_entry_above_zero(qa, sym):
    chk(qa, f"bot {sym} entry > 0", lambda: ((lambda p: (p["entryPrice"] > 0, U.show_price(p["entryPrice"])))(qa.dex.get_position(qa.who["alice"], MID[sym]))))


@then("the bot's free collateral fell")
def bot_free_fell(qa):
    def ev():
        f = qa.dex.free_collateral(qa.who["alice"])
        bal = qa.dex.vault_balance(qa.who["alice"])
        return f < bal, f"free ${U.show_usd(f)} < balance ${U.show_usd(bal)}"
    chk(qa, "bot free collateral fell", ev)


@then("the treasury's vault balance increased")
def treasury_increased(qa):
    chk(qa, "treasury balance increased", lambda: ((lambda b: (b > qa.noted["bal.treasury"], f"before {qa.noted['bal.treasury']}, after {b}"))(qa.dex.vault_balance(qa.who["treasury"]))))


@then("the bot's fill was mined")
def fill_mined(qa):
    def ev():
        result = (qa.bot_last or {}).get("receipt") or {}
        r = result.get("receipt") or {}
        return bool(r) and r.get("status") == "0x1" and bool(r.get("blockNumber")), ("status " + str(r.get("status")) + " block " + str(int(r["blockNumber"], 16)) if r else "no receipt")
    qa.observe("fill mined", ev)


@then(P("the account nonce advanced by {n:d}"))
def nonce_advanced(qa, n):
    def ev():
        now = int(rpc("eth_getTransactionCount", [qa.who["alice"], "latest"], qa.dex.url), 16)
        first = qa.bot.nonces[0]
        consecutive = all(v == qa.bot.nonces[i - 1] + 1 for i, v in enumerate(qa.bot.nonces) if i > 0)
        return consecutive and now - first == n, f"nonces {qa.bot.nonces}, now {now}"
    chk(qa, f"nonce advanced by {n}", ev)


@then(P("the bot's {sym:w} order clears the minimum notional"))
def clears_min_notional(qa, sym):
    def ev():
        o = qa.bot_last["order"]
        p = qa.dex.get_market_params(MID[sym])
        notional = U.notional(o["size"], o["price"])
        return notional >= p["minNotional"], f"notional ${U.show_usd(notional)} >= min ${U.show_usd(p['minNotional'])}"
    chk(qa, sym + " order clears min notional", ev)
